TOTP secrets in the web UI

mpflanzer opened the pull request:
Extends the UI to create, delete TOTP token generators and read TOTP tokens. The details view for an account refreshes automatically to generate the next token. 

The "provider" mode of the TOTP engine is currently not supported. This could easily be added in another PR if there is a use case for it.

Closes: #12698 and #16786

mpflanzer commented:
![image](https://github.com/hashicorp/vault/assets/520198/f6d02207-ac88-4b24-847d-271873976184)
![image](https://github.com/hashicorp/vault/assets/520198/ea969845-5839-4978-aa58-a1f7f8f5f6c5)
![image](https://github.com/hashicorp/vault/assets/520198/b7442e7e-9ed9-4804-932f-ebf80d4ccd51)
![image](https://github.com/hashicorp/vault/assets/520198/4332a640-70e4-4e3e-b4e9-f9c0bf2b55d1)
![image](https://github.com/hashicorp/vault/assets/520198/1f54bda0-67ab-4e05-a4b2-d7462fbf01b7)

mpflanzer commented:
I haven't added any test yet but can do if required. But it would be great to get a first review if the approach in general is fine.

hashishaw reviewed (changes requested):
Wow! 🤩 Amazing start on this, and kudos for navigating the notoriously difficult secrets area of the codebase! 

Overall, I think your approach for the TOTP keys makes sense, as it closely follows CRUD. However, for the code workflow I think we should copy a combination of AWS and Database:
- list view -- rename "Accounts" to "Keys" to better match documentation
- click on item takes you to /vault/secrets/totp/credentials/:keyName
- in the credentials route, if the backend type is `totp` we go ahead and get the credentials and show them on a credentials view, similar to how database credentials works
- From the keys list view, you can click the popup menu to access show or edit pages for the key details

For the code generation and verification, since the payload and responses are both fairly small it might be worth removing the `code-totp` model/adapter and using a custom adapter method for those endpoints instead. 

The "show" and "edit" views should show/edit the TOTP Key data, not the code. The code generation will be moved to the credentials page. 

Again, this is a great start and please feel free to tag @hashicorp/vault-ui if you have any questions!

mpflanzer commented:
Thanks for the detailed feedback. I'll try to find some time this weekend to make the changes according to your suggestions.

mpflanzer commented on ui/app/components/totp-create.js:
```diff
@@ -0,0 +1,159 @@
+  @service router;
+  @service store;
+
+  checkValidation(name, value) {
```
I added the `@withModelValidations` decorator but I'm not sure where to call the validate since `totp-edit.js` extends the `role-edit.js`. Could I just add the `validate()` call in the `createOrUpdate` action or would that break for models without the decorator? https://github.com/hashicorp/vault/blob/7f0e64480fb4aec87df92359559a5a16cdeee0cd/ui/app/components/role-edit.js#L69

mpflanzer commented on ui/app/models/totp.js:
```diff
@@ -0,0 +1,98 @@
+
+  @attr('string') url;
+
+  @computed('account_name', function () {
```
I'm not sure about this computed property. Now the `account_name` form field loses focus on every keystroke. Does `attrs` need to be a computed property? Or could it just be a getter?

hashishaw replied:
Right, this should be a getter! In fact all of the `@computed` values should be a getter instead.

mpflanzer commented:
I adapted the create/edit and generate workflow to your suggestions @hashishaw. I think I addressed most of your comments. Some open questions that I still have:

**To which extend does it make sense to allow editing the key entries?**
I don't see the use case to be able to edit keys for which Vault operates as generator. Maybe changing the account name makes sense but probably not the parameters for the generator. In provider mode editing could make sense but once the key is created the two modes cannot be distinguished anymore.

**How to deal with the URL vs individual fields to create a new key**
In generator mode keys can be created either with the `otpauth://` URL or by providing all the required information separately. In provider mode the URL option doesn't exists. One option would be parse the URL in the frontend, populate the individual fields with the extracted values and never send the `url` parameter to the backend. Or based on which fields are filled the request to the backend is build.

**Is the key name equal to the TOTP account_name?**
The TOTP model has an `account_name` attribute which typically would be used as the name for the key. If we want the same we need to at least extract the account name from the `otpauth://` URL chooses that option to create a new key.

**Support for provider mode?**
Do we want to support creating keys in provider mode at all? Doesn't make a big difference. Just after creating a new key the shared secret would need to be displayed. That step doesn't exist for the provider mode. Except for that the two modes cannot be distinguished later on (I think).

 @hashicorp/vault-ui

![image](https://github.com/hashicorp/vault/assets/520198/a97e6d4c-edb3-4caf-9366-9a7edbd9c2f1)

mpflanzer commented:
I added the provider mode as well. After the key created the user is presented with the QR code:
![image](https://github.com/hashicorp/vault/assets/520198/bad9d499-8865-47db-a4b3-3901ce4139d3)

mpflanzer commented:
The create dialog is still not great because of the ambiguity between the URL field and the other fields.
![image](https://github.com/hashicorp/vault/assets/520198/6906822a-b74f-41b7-a9ea-1ebdb4a8667c)

mpflanzer commented on ui/app/models/totp.js:
```diff
@@ -0,0 +1,171 @@
+  @attr('string', {
+    editDisabled: true,
+    possibleValues: ALGORITHMS,
+    defaultValue: 'SHA1',
```
Is it possible to provide default values for the create/edit form without having to set them here on the model?

hashishaw replied:
Sort of -- do you want the default value to be dynamic, or why isn't setting it on the model preferable?

mpflanzer replied:
No, the main reason was that everything which is set on the model is currently serialized. But some options are mutually exclusive in the `POST` request. I just realized that it's probably better to tackle this in the serializer and conditionally remove attributes from the serialized data.
The other reason was that some attributes are only used in the `POST` request but never returned by the `GET` request. But with the default values it looks like they got returned from the request as well.

hashishaw replied:
Got it! Yes, we don't want to show "default" values that are not actually reflected in the backend. One way we get around this limitation (which we run into fairly often) is to instead set the "initial values" in the `createRecord` invocation. So in this case, in the credentials route, we would have: 
```
getTOTPCode(backend, keyName) {
    return this.store.adapterFor('totp').generateCode(backend, keyName, { algorithm: 'SHA1' });
  },
```
with whatever other values we want to set initially, and remove the `defaultValue` unless this is truly the default value that the backend has. Good question!

hashishaw commented on ui/app/models/totp.js:
```diff
@@ -0,0 +1,171 @@
+      });
+    }
+
+    return fieldToAttrs(this, groups);
```
Sometimes we find that having dynamic fields doesn't play very nice with `fieldToAttrs` -- our newer pattern has been to add the `@withExpandedAttributes` decorator to the model, and then here you could replace with 
```suggestion
    return this._expandGroups(groups);
```

hashishaw commented on ui/app/models/totp.js:
```diff
@@ -0,0 +1,171 @@
+
+  @computed('url', 'barcode', function () {
+    const keys = ['url'];
+    return expandAttributeMeta(this, keys);
```
If you do use `withExpandedAttributes` on this model, `expandAttributeMeta` can be replaced with something like 
```
return keys.map(k => this.allByKey[k]);
```
We do this so that the expanded attribute meta isn't recalculated, just once on init

hashishaw commented on ui/app/routes/vault/cluster/secrets/backend/credentials.js:
```diff
@@ -64,15 +67,23 @@ export default Route.extend({
       awsRole = await this.getAwsRole(backendPath, role);
+    } else if (backendType === 'totp') {
+      totpCode = await this.getTOTPCode(backendPath, role);
+      const totpKey = await this.store.queryRecord('totp', { id: role, backend: backendPath });
```
Just flagging that doing it this way means a user with only generate permissions in their policy will get a failure on this page because it will also attempt to read the key. If we wanted to make the period optional you could do something similar to the `getAwsRole` method to fail silently on the non-essential data

hashishaw commented on ui/app/templates/components/generate-credentials-totp.hbs:
```diff
@@ -0,0 +1,51 @@
+      @color="secondary"
+      @route="vault.cluster.secrets.backend.list-root"
+      @model={{this.backendPath}}
+      data-test-secret-generate-back={{true}}
```
nit -- we don't need =true on the tests selectors anymore (that is leftover from earlier versions of Ember when we did)
```suggestion
      data-test-secret-generate-back
```

hashishaw commented on ui/app/templates/components/totp-edit.hbs:
```diff
@@ -0,0 +1,137 @@
+    />
+  </p.top>
+  <p.levelLeft>
+    <h1 class="title is-3" data-test-secret-header="true">
```
```suggestion
    <h1 class="title is-3" data-test-secret-header>
```

hashishaw commented on ui/app/templates/components/totp-edit.hbs:
```diff
@@ -0,0 +1,137 @@
+        {{#if (eq this.mode "create")}}
+          {{#each this.model.createAttrs as |attr|}}
+            <FormField
+              data-test-field={{true}}
```
FormField already has this selector so no need to provide it
```suggestion
```

hashishaw commented:
Thank you, this is excellent progress! Pardon me while I get up to speed on TOTP 😄 

> **To which extend does it make sense to allow editing the key entries?** I don't see the use case to be able to edit keys for which Vault operates as generator. Maybe changing the account name makes sense but probably not the parameters for the generator. In provider mode editing could make sense but once the key is created the two modes cannot be distinguished anymore.

Good question -- let me take this back to my team for discussion. I don't think we have any other engines that don't allow edit at all, and according to the API it looks like any of the fields other than name *could* be updated (whether that's advisable is something different). 

> **How to deal with the URL vs individual fields to create a new key** In generator mode keys can be created either with the `otpauth://` URL or by providing all the required information separately. In provider mode the URL option doesn't exists. One option would be parse the URL in the frontend, populate the individual fields with the extracted values and never send the `url` parameter to the backend. Or based on which fields are filled the request to the backend is build.

Another great question. Generally we prefer not to mess too much with the user input, and submit as-is. 

For displaying the fields relevant to the mode, we do have some patterns for choosing a type of setup that maybe we could incorporate here. I will talk to our design team, but relying on the user to know what to fill out might be fine for now. Do you know, if you create a key and set `url` only, does the read response only return with the url value? 

> **Is the key name equal to the TOTP account_name?** The TOTP model has an `account_name` attribute which typically would be used as the name for the key. If we want the same we need to at least extract the account name from the `otpauth://` URL chooses that option to create a new key.

It looks like the key's name is different than `account_name` according to the [API docs](https://developer.hashicorp.com/vault/api-docs/secret/totp#create-key). Are you saying that in practice, they are usually the same? 

> **Support for provider mode?** Do we want to support creating keys in provider mode at all? Doesn't make a big difference. Just after creating a new key the shared secret would need to be displayed. That step doesn't exist for the provider mode. Except for that the two modes cannot be distinguished later on (I think).

Your solution for provider mode seems reasonable, and I think it's good to support it with this work. If the key read response returns `generate: true` when that flag is set on create then we could distinguish later, but the docs don't seem to be explicit about that detail. 

Thanks again for your work on this! I will pull down the latest changes and test, and let you know if there are any other things we should update.

mpflanzer commented:
> Thank you, this is excellent progress! Pardon me while I get up to speed on TOTP 😄
No worries, most of the details were new to me as well. No need to rush on this. I'll anyway mostly have time on the weekends to address the comments.

> > **To which extend does it make sense to allow editing the key entries?** I don't see the use case to be able to edit keys for which Vault operates as generator. Maybe changing the account name makes sense but probably not the parameters for the generator. In provider mode editing could make sense but once the key is created the two modes cannot be distinguished anymore.
> 
> Good question -- let me take this back to my team for discussion. I don't think we have any other engines that don't allow edit at all, and according to the API it looks like any of the fields other than name _could_ be updated (whether that's advisable is something different).

I just tried to edit the `issuer` of a key but that failed with an error about missing information. Also by looking at the backend code (https://github.com/hashicorp/vault/blob/9af5c5c93364bb3fa7ea0f8a25351a53df49f4f8/builtin/logical/totp/path_keys.go#L210) I think there is no real editing. I can just completely replace all the information under a given key name but not edit part of it. In HTTP terms I would describe it as a PUT but not a PATCH API. The problem is that the TOTP secret is not known by the frontend so it cannot send it along with the request to edit. I have the feeling that doesn't play nicely with how editing usually works for the secrets in Vault. If I didn't overlook something I would recommend to disable editing to avoid confusion about what it's doing. Especially since a "delete and create" would achieve the same.
 
> > **How to deal with the URL vs individual fields to create a new key** In generator mode keys can be created either with the `otpauth://` URL or by providing all the required information separately. In provider mode the URL option doesn't exists. One option would be parse the URL in the frontend, populate the individual fields with the extracted values and never send the `url` parameter to the backend. Or based on which fields are filled the request to the backend is build.
> 
> Another great question. Generally we prefer not to mess too much with the user input, and submit as-is.
> 
> For displaying the fields relevant to the mode, we do have some patterns for choosing a type of setup that maybe we could incorporate here. I will talk to our design team, but relying on the user to know what to fill out might be fine for now. Do you know, if you create a key and set `url` only, does the read response only return with the url value?

I'm not sure which exact request you're talking about. So I'll just provide the answer for all of them. :wink:
- In provider mode (generate=`true`) the `POST` request to generate a key returns `url` and optionally `barcode`.
- In generator mode (generate=`false`) the `POST` request to generate a key returns nothing
- Independent of the mode that was used to generate the key the `GET` request to read the details for a key never returns the `url` as this would contain the secret. Only basic information like `account_name`, `alogrithm`, `digits` are reported.

If generate is `false` if a URL is provided all information contained in it takes precedence over the other fields. But things missing from an incomplete URL will be used from the fields. So we could just say we have the field for the URL and all other fields and maybe add a short description for the user how to fill it.
 
> > **Is the key name equal to the TOTP account_name?** The TOTP model has an `account_name` attribute which typically would be used as the name for the key. If we want the same we need to at least extract the account name from the `otpauth://` URL chooses that option to create a new key.
> 
> It looks like the key's name is different than `account_name` according to the [API docs](https://developer.hashicorp.com/vault/api-docs/secret/totp#create-key). Are you saying that in practice, they are usually the same?

Yes, I would say so. For example in the screenshot from FreeOTP the accounts are just identified by information like account name and issuer. I'm not sure if you could add a custom key name as well or edit some of the information.
I don't have a very strong opinion of how we should do it. The question is mostly for me to know how I should implement it. I could imagine the following scenarios:

- A separate and mandatory field for the key name
- A separate but optional field for the key name. If the user doesn't fill it the information from the other fields is used.
- No extra field for the key name and it is always derived from the other fields.

The last two options might come with some issues to transform the values into a valid key name. I'm not sure if all characters are allowed as part of the key name. And if we allow to edit the `account_name` later it would anyway diverge from the key name as that wouldn't change I assume.

> > **Support for provider mode?** Do we want to support creating keys in provider mode at all? Doesn't make a big difference. Just after creating a new key the shared secret would need to be displayed. That step doesn't exist for the provider mode. Except for that the two modes cannot be distinguished later on (I think).
> 
> Your solution for provider mode seems reasonable, and I think it's good to support it with this work. If the key read response returns `generate: true` when that flag is set on create then we could distinguish later, but the docs don't seem to be explicit about that detail.

Unfortunately that information is not stored. From a TOTP perspective it matters only during the setup and doesn't make any difference later which is I guess why it is not stored. But as you said, I think we can stick to what I currently implemented and that should be fine.

> Thanks again for your work on this! I will pull down the latest changes and test, and let you know if there are any other things we should update.

You're welcome, thanks for the review and feedback. I'll address your new comments over the next days.

mpflanzer commented:
@hashishaw do you have any update on this?

e-scheer commented:
any ETA regarding that feature ?

hyzza commented:
thanks for working on this, so important feature for us..any eta on release?

mpflanzer commented:
Is there still interest in merging this into Vault?

hyzza commented:
yes

hellobontempo commented:
Thank you all for showing continued interest with this issue. We are discussing this internally - I'm sorry it's taking longer than initially anticipated.

sagianael commented:
> Is there still interest in merging this into Vault?

Yes

hellobontempo commented:
Thank you for your work on this @mpflanzer ! We're merging this into a sidebranch so we can make some small UX improvements based on feedback from our design team. In addition, there are some updated Ember patterns we'd like to implement here. Again, thank you for your contribution! We hope to release this in 1.20. I will circle back and apply a milestone when that is confirmed (that should be in a couple weeks).
